pkgs/checks/test/context_test.dart

```diff
@@ -0,0 +1,115 @@
+
+import 'package:async/async.dart' hide Result;
+import 'package:checks/checks.dart';
+import 'package:checks/context.dart';
+import 'package:test/scaffolding.dart';
+import 'package:test_api/hooks.dart';
+import 'package:test_api/hooks_testing.dart';
+
+void main() {
+  group('Context', () {
+    test('expectAsync holds test open', () async {
+      late void Function() callback;
+      final monitor = TestCaseMonitor.start(() {
+        check(null).context.expectAsync(() => [''], (actual) async {
+          final completer = Completer<void>();
+          callback = completer.complete;
+          await completer.future;
+          return null;
+        });
+      });
+      await pumpEventQueue();
+      check(monitor).status.equals(Status.running);
+      callback();
+      await monitor.onDone;
+      check(monitor).didPass();
+    });
+
+    test('nestAsync holds test open', () async {
+      late void Function() callback;
+      final monitor = TestCaseMonitor.start(() {
+        check(null).context.nestAsync(() => [''], (actual) async {
+          final completer = Completer<void>();
+          callback = completer.complete;
+          await completer.future;
+          return Extracted.value(null);
+        }, it());
+      });
+      await pumpEventQueue();
+      check(monitor).status.equals(Status.running);
+      callback();
+      await monitor.onDone;
+      check(monitor).didPass();
+    });
+
+    test('expectUnawaited can fail the test after it completes', () async {
+      late void Function() callback;
+      final monitor = await TestCaseMonitor.run(() {
+        check(null).context.expectUnawaited(() => [''], (actual, reject) {
+          final completer = Completer<void>()
+            ..future.then((_) {
+              reject(Rejection(which: ['']));
+            });
+          callback = completer.complete;
+        });
+      });
+      check(monitor)
+        ..status.equals(Status.complete)
+        ..result.equals(Result.success);
+      callback();
+      await pumpEventQueue();
+      check(monitor)
+          .didFail(Result.error)
+          .contains('This test failed after it had already completed.');
+    });
+  });
+}
+
+extension _MonitorChecks on Subject<TestCaseMonitor> {
+  Subject<Status> get status => has((m) => m.status, 'status');
+  Subject<Result> get result => has((m) => m.result, 'result');
+  Subject<Iterable<AsyncError>> get errors => has((m) => m.errors, 'errors');
+  Subject<StreamQueue<AsyncError>> get onError =>
+      has((m) => m.onError, 'onError').withQueue;
+
+  /// Expects that the monitored test is completed as success with no errors.
+  ///
+  /// Sets up an unawaited expectation that the test does not emit errors in the
+  /// future in addition to checking there have been no errors yet.
+  void didPass() {
+    errors.isEmpty();
+    status.equals(Status.complete);
+    result.equals(Result.success);
+    onError.context.expectUnawaited(() => ['emits no further errors'],
```

Comment:
`neverEmits` expects the stream to close.

I wonder if we should change that and have you write `check(stream).allOf([it()..neverEmits(condition), it()..isDone()]);`

Reply:
I think `neverEmits` should expect the stream to close, otherwise how long should it wait for? If you want to fail asynchronously then `expectUnawaited` is there for you.

Reply:
Yeah, the case for expecting the stream to close is more common. I don't like having to use the context directly like this, and having the expectations be orthogonal is kind of nice, but there is a strong argument for making the very common case the only implementation we provide. I'll definitely wait for a separate pr if we make any change around this, was just taking a note here that the API did cause a problem at this usage site.